storage/src/main/java/org/apache/kafka/server/log/internals/LastRecord.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
+
+import java.util.Objects;
+import java.util.OptionalLong;
+
+/**
+ * The last written record for a given producer. The last data offset may be undefined
+ * if the only log entry for a producer is a transaction marker.
+ */
+public final class LastRecord {
+    public final OptionalLong lastDataOffset;
+    public final short producerEpoch;
+
+    public LastRecord(OptionalLong lastDataOffset, short producerEpoch) {
+        this.lastDataOffset = lastDataOffset;
+        this.producerEpoch = producerEpoch;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        LastRecord that = (LastRecord) o;
+
+        if (producerEpoch != that.producerEpoch) return false;
+        return Objects.equals(lastDataOffset, that.lastDataOffset);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = lastDataOffset != null ? lastDataOffset.hashCode() : 0;
+        result = 31 * result + (int) producerEpoch;
```

Comment:
Is this cast actually needed?

Reply:
Good catch! It is generated by intellij. I do not think this is really needed as you suggested.